google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,469 @@
+
+/**
+ * `ResumableAsyncStreamingReadWriteRpc<ResponseType, RequestType>` uses
+ * callables compatible with this `std::function<>` to initialize a stream
+ * from AsyncStreamFactory.
+ */
+template <typename RequestType, typename ResponseType>
+using StreamInitializer = std::function<future<StatusOr<
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>>>(
+    std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>)>;
+
+using AsyncSleeper = std::function<future<void>(std::chrono::duration)>;
+
+template <typename RequestType, typename ResponseType>
+class ResumableAsyncStreamingReadWriteRpc {
+ public:
+  virtual ~ResumableAsyncStreamingReadWriteRpc() = default;
+
+  /**
+   * Start the streaming RPC.
+   *
+   * The returned future will be returned with a status
+   * when this stream will not be resumed or when the user calls `Finish()`.
+   */
+  virtual future<Status> Start() = 0;
+
+  /**
+   * Read one response from the streaming RPC.
+   *
+   * @note Only **one** `Read()` operation may be pending at a time. The
+   * application is responsible for waiting until any previous `Read()`
+   * operations have completed before calling `Read()` again.
+   *
+   * Whether `Read()` can be called before a `Write()` operation is specified by
+   * each service and RPC. Most services require at least one `Write()` call
+   * before calling `Read()`.  Many services may return more than one response
+   * for a single `Write()` request.  Each service and RPC specifies how to
+   * discover if more responses will be forthcoming.
+   *
+   * If the `optional<>` is not engaged, a successful `ResponseType` is
+   * returned. If it is engaged, the call failed, but the user may call `Read`
+   * again unless `Start` had finished with a permanent error `Status`. The
+   * application should wait until any other pending operations (typically any
+   * other `Write()` calls) complete and then call `Finish()` to find the status
+   * of the streaming RPC.
+   */
+  virtual future<absl::optional<ResponseType>> Read() = 0;
+
+  /**
+   * Write one request to the streaming RPC.
+   *
+   * @note Only **one** `Write()` operation may be pending at a time. The
+   * application is responsible for waiting until any previous `Write()`
+   * operations have completed before calling `Write()` again.
+   *
+   * Whether `Write()` can be called before waiting for a matching `Read()`
+   * operation is specified by each service and RPC. Many services tolerate
+   * multiple `Write()` calls before performing or at least receiving a `Read()`
+   * response.
+   *
+   * If `true` is returned, the call was successful. If `false` is returned,
+   * the call failed, but the user may call `Write` again unless `Start` had
+   * finished with a permanent error `Status`. The application should wait until
+   * any other pending operations (typically any other `Read()` calls) complete
```

Comment:
Probably remove this?

Reply:
the whole 'If `true` ...` paragraph?

Reply:
no, just the "The application should wait" part? I guess it's still necessary